Add irregular txType, add check for total balance, prevent proposal withhold attack

ManfredKarrer opened the pull request.

ManfredKarrer committed:
Don't create vote reveal tx in last block of phase

ManfredKarrer committed:
Add IRREGULAR txType for txs which are not rule conform but not burnt

We don't want to burn BSQ in cases like that the tx was published too
late, which is a valid case if the tx does not make it in the next block.
We set such txs as IRREGULAR and allow spending of the BSQ, but there
function in the governance is invalidated.

We also add a check if the sum of all UTXO is the same as the sum of the
genesis + sum of issuance txs - burned fees.

ManfredKarrer committed:
Merge branch 'master' into add-check-for-bsq-balance

# Conflicts:
#	desktop/src/main/java/bisq/desktop/main/dao/wallet/dashboard/BsqDashboardView.java

sqrrm commented on core/src/main/java/bisq/core/dao/governance/votereveal/VoteRevealService.java:
```diff
@@ -179,8 +179,11 @@ private void maybeRevealVotes(int chainHeight) {
                     boolean isInVoteRevealPhase = periodService.getPhaseForHeight(chainHeight) == DaoPhase.Phase.VOTE_REVEAL;
+                    // If we would create the tx in the last block it would be confirmed in the best case in th next
+                    // block which would be already the break and would invalidate the vote reveal.
+                    boolean isNotLastBlockInPhase = chainHeight != periodService.getLastBlockOfPhase(chainHeight, DaoPhase.Phase.VOTE_REVEAL);
```
Would be easier to understand the code if the name was isLastBlockInPhase and the check later was for !isLastBlockInPhase.

ManfredKarrer replied:
Ok, changed it

sqrrm commented on core/src/main/java/bisq/core/dao/governance/votereveal/VoteRevealService.java:
```diff
@@ -179,8 +179,11 @@ private void maybeRevealVotes(int chainHeight) {
                     boolean isInVoteRevealPhase = periodService.getPhaseForHeight(chainHeight) == DaoPhase.Phase.VOTE_REVEAL;
+                    // If we would create the tx in the last block it would be confirmed in the best case in th next
+                    // block which would be already the break and would invalidate the vote reveal.
+                    boolean isNotLastBlockInPhase = chainHeight != periodService.getLastBlockOfPhase(chainHeight, DaoPhase.Phase.VOTE_REVEAL);
```
```suggestion
                    boolean isLastBlockInPhase = chainHeight == periodService.getLastBlockOfPhase(chainHeight, DaoPhase.Phase.VOTE_REVEAL);
```

sqrrm commented on core/src/main/java/bisq/core/dao/governance/votereveal/VoteRevealService.java:
```diff
@@ -179,8 +179,11 @@ private void maybeRevealVotes(int chainHeight) {
+                    boolean isNotLastBlockInPhase = chainHeight != periodService.getLastBlockOfPhase(chainHeight, DaoPhase.Phase.VOTE_REVEAL);
                     boolean isBlindVoteTxInCorrectPhaseAndCycle = periodService.isTxInPhaseAndCycle(myVote.getTxId(), DaoPhase.Phase.BLIND_VOTE, chainHeight);
-                    if (isInVoteRevealPhase && isBlindVoteTxInCorrectPhaseAndCycle) {
+                    if (isInVoteRevealPhase && isNotLastBlockInPhase && isBlindVoteTxInCorrectPhaseAndCycle) {
```
```suggestion
                    if (isInVoteRevealPhase && !isLastBlockInPhase && isBlindVoteTxInCorrectPhaseAndCycle) {
```

sqrrm commented on core/src/main/java/bisq/core/dao/node/parser/TxParser.java:
```diff
@@ -249,17 +256,17 @@ private void processIssuance(int blockHeight, TempTx tempTx, long bsqFee) {
+            tempTx.setTxType(TxType.IRREGULAR);
 
-            // TODO don't burn in such cases
+            // Make sure the optionalIssuanceCandidate is set the BTC
```
```suggestion
            // Make sure the optionalIssuanceCandidate is set to BTC
```

sqrrm commented on core/src/main/java/bisq/core/dao/node/parser/TxParser.java:
```diff
@@ -393,19 +399,19 @@ static TxType evaluateTxTypeFromOpReturnType(TempTx tempTx, OpReturnType opRetur
-                    return TxType.INVALID;
+                    // This is not a valid issuance tx
+                    // We tolerate such an incorrect tx and do not burn the BSQ
+                    return TxType.IRREGULAR;
```
In this case I would also be ok to burn the BSQ. This is a malformed tx which should not be possible to create with an unmodified client. This is one of the cases where being strict might be worth it as this is about issuance.

ManfredKarrer replied:
Agree.

sqrrm commented on core/src/main/java/bisq/core/dao/node/parser/TxParser.java:
```diff
@@ -393,19 +399,19 @@ static TxType evaluateTxTypeFromOpReturnType(TempTx tempTx, OpReturnType opRetur
-                    return TxType.INVALID;
+                    // This is not a valid issuance tx
+                    // We tolerate such an incorrect tx and do not burn the BSQ
+                    return TxType.IRREGULAR;
```
Same here about being strict. Possibly worth being stricter unless you can see a way that normal usage of the client would result in this kind of malformed issuance tx.

ManfredKarrer replied:
Agree. But changing it later will be likely not possible (see mess it has created with currnet change).

sqrrm commented on core/src/main/java/bisq/core/dao/node/parser/TxParser.java:
```diff
@@ -423,8 +429,8 @@ static TxType evaluateTxTypeFromOpReturnType(TempTx tempTx, OpReturnType opRetur
-                // TODO do we want to burn in such a case?
-                return TxType.INVALID;
+                // We tolerate such an incorrect tx and do not burn the BSQ
+                return TxType.IRREGULAR;
```
This case I'm not sure about. Perhaps it's good to be this lenient as this is no special kind of tx and would just result in a normal BSQ tx sending BSQ from one output to the next.

ManfredKarrer replied:
Ok.

sqrrm commented on desktop/src/main/java/bisq/desktop/main/dao/wallet/dashboard/BsqDashboardView.java:
```diff
@@ -199,18 +200,25 @@ private void updateWithBsqBlockChainData() {
-        availableAmount = issuedAmountFromGenesis
+        Coin totalUtxoAmount = Coin.valueOf(daoFacade.getTotalAmountOfUnspentTxOutputs());
+
+        daoFacade.getTotalAmountOfConfiscatedTxOutputs();
```
```suggestion
```

sqrrm reviewed (approved):
utACK

I have some suggested changes that you can use if you feel like.

There is also some questions on how lenient we should be with irregular txs for failed issuance, just added my thoughts.

sqrrm committed:
Update core/src/main/java/bisq/core/dao/governance/votereveal/VoteRevealService.java

Co-Authored-By: ManfredKarrer <mk@nucleo.io>

sqrrm committed:
Update core/src/main/java/bisq/core/dao/governance/votereveal/VoteRevealService.java

Co-Authored-By: ManfredKarrer <mk@nucleo.io>

sqrrm committed:
Update core/src/main/java/bisq/core/dao/node/parser/TxParser.java

Co-Authored-By: ManfredKarrer <mk@nucleo.io>

sqrrm committed:
Update desktop/src/main/java/bisq/desktop/main/dao/wallet/dashboard/BsqDashboardView.java

Co-Authored-By: ManfredKarrer <mk@nucleo.io>

ManfredKarrer committed:
Keep invalid state for txs which cannot have been created by the app

ManfredKarrer committed:
Merge branch 'add-check-for-bsq-balance' of https://github.com/ManfredKarrer/bisq into add-check-for-bsq-balance

ManfredKarrer committed:
Rename BTC_DAO_TESTNET to BTC_DAO_TESTNET2 to enforce new network

ManfredKarrer committed:
Add try/catch to print error in case the file does not exist

ManfredKarrer committed:
Rename BTC_DAO_TESTNET2 to BTC_DAO_REGTEST

ManfredKarrer committed:
Add new genesis tx for DAO_REGTEST

ManfredKarrer committed:
Reduce durations of phases for DAO_REGTEST

ManfredKarrer committed:
use new DAO_REGTEST server and seeds

ManfredKarrer committed:
Fix incorrect seed address

ManfredKarrer committed:
Add log for seeds

sqrrm commented on core/src/main/java/bisq/core/dao/governance/param/Param.java:
```diff
@@ -142,55 +142,55 @@
                     BisqEnvironment.getBaseCurrencyNetwork().isDaoBetaNet() ?
                             "144" :       // daoBetaNet; 1 day
-                            "380",      // testnet or dao testnet (server side regtest); 2.6 days
+                            "134",      // testnet or dao regtest (server side regtest); 2.6 days
```
```suggestion
                            "134",      // testnet or dao regtest (server side regtest); 0.93 days
```

sqrrm commented on core/src/main/java/bisq/core/dao/governance/param/Param.java:
```diff
@@ -142,55 +142,55 @@
                     BisqEnvironment.getBaseCurrencyNetwork().isDaoBetaNet() ?
                             "144" :       // daoBetaNet; 1 day
-                            "300",      // testnet or dao testnet (server side regtest); 2 days
+                            "134",      // testnet or dao regtest (server side regtest); 2 days
```
```suggestion
                            "134",      // testnet or dao regtest (server side regtest); 0.93 days
```

sqrrm commented on core/src/main/java/bisq/core/dao/governance/param/Param.java:
```diff
@@ -142,55 +142,55 @@
                     BisqEnvironment.getBaseCurrencyNetwork().isDaoBetaNet() ?
                             "144" :       // daoBetaNet; 1 day
-                            "300",      // testnet or dao testnet (server side regtest); 2 days
+                            "132",      // testnet or dao regtest (server side regtest); 2 days
```
```suggestion
                            "132",      // testnet or dao regtest (server side regtest); 0.93 days
```

sqrrm reviewed (approved):
utACK

ManfredKarrer committed:
Merge branch 'master' into add-check-for-bsq-balance

# Conflicts:
#	core/src/main/resources/i18n/displayStrings.properties
#	desktop/src/main/java/bisq/desktop/main/dao/wallet/dashboard/BsqDashboardView.java

ManfredKarrer committed:
Remove unused param

sqrrm committed:
Update core/src/main/java/bisq/core/dao/governance/param/Param.java

Co-Authored-By: ManfredKarrer <mk@nucleo.io>

sqrrm committed:
Update core/src/main/java/bisq/core/dao/governance/param/Param.java

Co-Authored-By: ManfredKarrer <mk@nucleo.io>

sqrrm committed:
Update core/src/main/java/bisq/core/dao/governance/param/Param.java

Co-Authored-By: ManfredKarrer <mk@nucleo.io>

ManfredKarrer committed:
Improve formatting

ManfredKarrer committed:
Cleanup, add string

ManfredKarrer committed:
Protect against proposal withhold attack

Set proposals which have been not included in the blind vote but which
have been later published and are part if the cycles proposals to
rejected.
A malicious voter could manipulate the software to withhold publishing
of his proposal and be the only voter on it. At the last block in the
blind vote phase he could publish the proposal but others cannot vote
anymore as they likely have already voted. In the vote result the
other voters would have treated it like ignored and if the voter had
enough BSQ to pass the quorum he could get accepted his proposal.
With this change we set all proposals which are not part in the blind
vote data but found in the cycle's ballot list as rejected.

ManfredKarrer commented:
Sorry again for pumping more in that PR. But as it will change data, hashchain, genesis tx... I prefer to have that all in one merge.

sqrrm commented on core/src/main/java/bisq/core/dao/governance/voteresult/VoteResultService.java:
```diff
@@ -358,29 +358,46 @@ private DecryptedBallotsWithMerits getDecryptedBallotsWithMerits(
+                                log.warn("We found a local vote but don't have that vote in the data from the " +
+                                        "blind vote. ballot={}", ballot);
+                            }
+                            if (ballot.getVote() != entry.getValue().get()) {
```
This needs to be an `else if`, `entry.getValue()` is checked in the previous `if`but code execution continues.

ManfredKarrer replied:
Yes will fix it.

sqrrm commented on core/src/main/java/bisq/core/dao/governance/voteresult/VoteResultService.java:
```diff
@@ -358,29 +358,46 @@ private DecryptedBallotsWithMerits getDecryptedBallotsWithMerits(
+                            }
+                        }
+
+                        // We only return accpeted or rejected votes
```
```suggestion
                        // We only return accepted or rejected votes
```

sqrrm commented on core/src/main/java/bisq/core/dao/governance/voteresult/VoteResultService.java:
```diff
@@ -396,6 +413,17 @@ private BallotList createBallotList(VoteWithProposalTxIdList voteWithProposalTxI
         if (!missingBallots.isEmpty())
             throw new VoteResultException.MissingBallotException(ballots, missingBallots);
 
+        // If we received a proposal after we had already voted we consider it as an proposla withhold attack and
```
```suggestion
        // If we received a proposal after we had already voted we consider it as an proposal withhold attack and
```

sqrrm committed:
Update core/src/main/java/bisq/core/dao/governance/voteresult/VoteResultService.java

Co-Authored-By: ManfredKarrer <mk@nucleo.io>

sqrrm committed:
Update core/src/main/java/bisq/core/dao/governance/voteresult/VoteResultService.java

Co-Authored-By: ManfredKarrer <mk@nucleo.io>

ManfredKarrer committed:
Fix if else case

ManfredKarrer committed:
Merge branch 'add-check-for-bsq-balance' of https://github.com/ManfredKarrer/bisq into add-check-for-bsq-balance

ManfredKarrer committed:
Merge branch 'master' into add-check-for-bsq-balance

# Conflicts:
#	core/src/main/resources/i18n/displayStrings.properties

ManfredKarrer committed:
Use lockupTxId instead of uid

sqrrm commented:
utACK

Took me a while to get the reject vote on proposals already seen locally but not included in a vote. It's clear that it does that, the comments are correct but it was still hard to follow. Most likely some detailed documentation of how the blindvote/payload/votereveal/voteresult process works is needed to be able to follow why all this happens and is needed.

ManfredKarrer committed:
Add random delay for proposalPayload publishing

We want to avoid that all nodes publish at the same time all proposals,
so we add a random delay from 100 ms - 5 sec.
A more sophisticated protection would be probably good but that delay
should help to avoid network spikes and is simple enough to not add
risks that the publishing would fail.

ManfredKarrer committed:
Cleanup

ManfredKarrer committed:
Remove setFitToRowsForTableView, update merit at activate

sqrrm commented on core/src/main/java/bisq/core/dao/governance/proposal/ProposalService.java:
```diff
@@ -219,12 +222,14 @@ private void publishToAppendOnlyDataStore() {
+                                    proposalPayload.getProposal().getTxId());
+                        else
+                            log.warn("publishToAppendOnlyDataStore failed for proposal " + proposalPayload.getProposal());
+                    }, 100, 5000, TimeUnit.MILLISECONDS);
```
```suggestion
                    }, 1, 1800, TimeUnit.SECONDS);
```

ManfredKarrer replied:
We don't want to risk that the user shuts down the app in the meantime. The long we delay the higher the risk.

sqrrm replied:
There are a lot of clients and they will all be sending the same data more or less so it's not dependent on one single instance. If you think that's not as safe I'm ok with the 5s interval as well.

ManfredKarrer replied:
Yes, maybe we need a more sophisticated solution here... or just increase the delay?

ManfredKarrer replied:
Implemented as discussed. We set reBroadcast to false to avoid flooding. Other broadcast handling is less critical as user only publishes its own proposals/votes. But still some case where I added a longer random delay to avoid network stress.

sqrrm reviewed (approved):
utACK

Perhaps a longer delay might be better, this is still quite a burst, why not spread it over half an hour? BREAK1 is 149 blocks long. Either way is fine though.

ManfredKarrer committed:
Add support for displaying burned BSQ from invalid txs

ManfredKarrer committed:
Add popup in case the utxo balance does not match the bsq balance

sqrrm commented on core/src/main/java/bisq/core/dao/state/DaoStateService.java:
```diff
@@ -811,6 +816,27 @@ public long getTotalAmountOfConfiscatedTxOutputs() {
+                .sum();
+    }
+
+    public long getBurnedBsqOfInvalidTx(Tx tx) {
```
This is not specific to burned txs. A better name might be `getBsqInputValue(Tx tx)`

ManfredKarrer replied:
Hm... I consider it burned BSQ as well in contrast to burnedFee.

sqrrm replied:
Yes, my point was that the function can calculate the input value from any tx, not only invalid ones.

ManfredKarrer replied:
Ah ok. I am not sure if it is valid for normal txs as well as all special cases handled in the input parser are ignored here. It only checks the txoutput types. I think for the use case of invalid txs that should be good enough. A maybe more safe approach would be to add another field to tx like burnedFee for the burned BSQ from an invalid tx. Only at parse time we have all checks done and we could not do that later as well. Hm... not sure if its worth to add that field but might be more safe. In fact we need to add a comment to that method that this is not intended to be used otherwise as it is not a perfect check....

sqrrm replied:
Yeah, good point. If we can add the field and set in the parser it would be more correct and safer I think.

ManfredKarrer replied:
Implemented as discussed. We use the burntFee field in Tx but renamed it.

sqrrm commented on core/src/main/java/bisq/core/dao/monitoring/DaoStateMonitoringService.java:
```diff
@@ -142,13 +148,13 @@ public void onDaoStateChanged(Block block) {
 
         if (sumBsq != sumUtxo) {
-            throw new RuntimeException("There is a mismatch between the UTXO set and the DAO state. Please contact the Bisq devlopers.");
+            utxoMismatches.add(new UtxoMismatch(block.getHeight(), sumUtxo, sumBsq));
```
If we get a mismatch it seems that will stay mismatched in the future as well and we'll get a new `UtxoMismatch` added at every block after the first one. Maybe this is good but could be a bit annoying if there is a popup for every new block.

ManfredKarrer replied:
Yes. Need to improve that. Popups are handled in the UI to not repeat but still we don't want to add it at each block again. But anyway should not happen at all, is just to double check if we missed something. So I consider it more as a developer tool...

ManfredKarrer replied:
I leave that for now but will keep it on my todo list

sqrrm reviewed (approved):
utACK

Looks ok but feel free to rename the function I commented on.

ManfredKarrer committed:
Set MAINNET_GENESIS_TOTAL_SUPPLY of old mainnet genesis

ManfredKarrer committed:
Improve Merit handling

- We add a filter to the getMeritStake method to avoid an exception at
getWeightedMeritAmount which would set merits to 0 in case the issuance
height blind would be larger than the vote height. I saw that error log
but could not reproduce it afterwards. It basically protects that we do
not add the issuance of the current cycle to the merit used in the
blind vote of that cycle.

- We add 1 block in the getCurrentlyAvailableMerit method to get the
same merit value if the blind vote was included in the next block.

ManfredKarrer committed:
Use onParseBlockCompleteAfterBatchProcessing to avoid sequence issues

When using onParseBlockChainComplete and
onParseBlockCompleteAfterBatchProcessing the
onParseBlockCompleteAfterBatchProcessing is called earlier as
onParseBlockChainComplete which caused incorrect state update in the UI.

ManfredKarrer committed:
Add better comments, cleanup

ManfredKarrer committed:
Refactoring: Rename method

ManfredKarrer committed:
Refactoring: Rename method

ManfredKarrer committed:
Fix comment

ManfredKarrer committed:
Refactoring: Rename method

ManfredKarrer committed:
Handle merits better

We did not update the merit correctly in case there was no proposal
selected.

ManfredKarrer committed:
Improve handling fo p2p network data broadcasts

ManfredKarrer committed:
Use burnedBsq field in Tx for burnedFee and invalidatedBsq

In case of an invalid tx we burn all available BSQ input. We only know
that at parsing time. We renamed the burntFee field to burntBsq to make
it more generic and use it for the burnt fee in case if a normal tx and
as invalidatedBsq in case of an invalid tx.

sqrrm commented on core/src/main/java/bisq/core/dao/governance/merit/MeritConsensus.java:
```diff
@@ -54,28 +54,27 @@ public static MeritList decryptMeritList(byte[] encryptedMeritList, SecretKey se
                         return getWeightedMeritAmount(issuance.getAmount(),
                                 issuance.getChainHeight(),
-                                txChainHeight,
+                                blindVoteTxHeight,
```
This seems like a good change, more consistent calculations. It also counts the merit at the time of voting which is what the user would expect.

sqrrm reviewed (approved):
utACK

I think changing the name to burntBsq made it a lot easier to read and understand. Looks like a good solution.

ManfredKarrer committed:
Add string validations

ManfredKarrer committed:
Fix missing close handler and avoid nullPointer exception

ManfredKarrer committed:
Use null instead of empty string for txId

ManfredKarrer committed:
Add more validation

- Check max length of strings and byte arrays
- Check that tx ID has 64 chars
- Add ExtraDataMapValidator for validating extraDataMap fields

ManfredKarrer committed:
Add BTC_DAO_TESTNET again to keep supporting current dao testnet

- It is more safe to separate the BTC_DAO_TESTNET and BTC_DAO_REGTEST
by the network ID as that prevents on the P2P network layer that the
network could interconnect. We would have risked that we receive network
data from the other network as users would use the persisted peers for
connections.

ManfredKarrer committed:
Merge branch 'master' into add-check-for-bsq-balance

# Conflicts:
#	core/src/main/resources/i18n/displayStrings_de.properties
#	core/src/main/resources/i18n/displayStrings_el.properties
#	core/src/main/resources/i18n/displayStrings_es.properties
#	core/src/main/resources/i18n/displayStrings_fa.properties
#	core/src/main/resources/i18n/displayStrings_fr.properties
#	core/src/main/resources/i18n/displayStrings_hu.properties
#	core/src/main/resources/i18n/displayStrings_pt.properties
#	core/src/main/resources/i18n/displayStrings_ro.properties
#	core/src/main/resources/i18n/displayStrings_ru.properties
#	core/src/main/resources/i18n/displayStrings_sr.properties
#	core/src/main/resources/i18n/displayStrings_th.properties
#	core/src/main/resources/i18n/displayStrings_vi.properties
#	core/src/main/resources/i18n/displayStrings_zh.properties

ManfredKarrer committed:
Add BTC_DAO_TESTNET

ManfredKarrer committed:
Improve comments, cleanup

ManfredKarrer committed:
Cleanup

sqrrm commented on common/src/main/java/bisq/common/util/ExtraDataMapValidator.java:
```diff
@@ -0,0 +1,75 @@
+        return getValidatedExtraDataMap(extraDataMap, MAX_SIZE, MAX_KEY_LENGTH, MAX_VALUE_LENGTH);
+    }
+
+    public static Map<String, String> getValidatedExtraDataMap(@Nullable Map<String, String> extraDataMap, int maxSize, int maxKeyLength, int maxValueLength) {
```
Please try to break these long lines, it's harder to read when they're too long.

ManfredKarrer replied:
Done

sqrrm reviewed (approved):
utACK

ManfredKarrer committed:
Update comment

ManfredKarrer committed:
Apply ExtraDataMapValidator for all extraDataMap fields

ManfredKarrer committed:
Add number of irregular txs to UI

ManfredKarrer committed:
Break up lines

ManfredKarrer commented:
Last commits have been rather trivial, but you are welcome to review them post merge as well.

ManfredKarrer committed:
Update witness file

ManfredKarrer commented:
Build error is prob caused due cache issue. Local build works and witness has not changed. Pushed anotehr change to trigger re-build.

sqrrm commented:
Latest changes also look good.
